File: data/src/main/java/com/aconno/acnsensa/data/converter/DataStringConverter.java
package com.aconno.acnsensa.data.converter;

import com.aconno.acnsensa.domain.model.readings.Reading;

import java.util.HashMap;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class DataStringConverter {

    private HashMap<String, String> map;

    private Pattern pattern = Pattern.compile("\\$\\s*(\\w+)");

    public DataStringConverter(String userDataString) {
        parseDataString(userDataString);
    }

    private void parseDataString(String everything) {
        map = new HashMap<>();

        Scanner scanner = new Scanner(everything);
        while (scanner.hasNextLine()) {
            String line = scanner.nextLine();

            Matcher matcher = pattern.matcher(line);
            while (matcher.find()) {
                if (!matcher.group(1).equalsIgnoreCase("ts")
                        && !matcher.group(1).equalsIgnoreCase("name")) {
                    map.put(matcher.group(1), line);
                }
            }
        }
        scanner.close();
    }

//    public String convert(Reading data) {
//        String dataString = map.get(data.getSensorType().toString().toLowerCase());
//        return dataString.replace("$" + data.getSensorType().toString().toLowerCase(), data.value)
//                .replace("$ts", String.valueOf(System.currentTimeMillis()))
//                .replace("$name", data.name);
//    }
}

File: data/src/main/java/com/aconno/acnsensa/data/converter/DataStringConverter.kt
package com.aconno.acnsensa.data.converter

import com.aconno.acnsensa.domain.model.Reading
import java.util.*
import java.util.regex.Pattern

class DataStringConverter(userDataString: String) {

    companion object {
        private const val VALUE = "value"
    }

    private var map: HashMap<String, String>? = null

    private val pattern = Pattern.compile("\\$\\s*(\\w+)")

    init {
        parseDataString(userDataString)
    }

    private fun parseDataString(everything: String) {
        map = HashMap()

        val scanner = Scanner(everything)
        while (scanner.hasNextLine()) {
            val line = scanner.nextLine()

            val matcher = pattern.matcher(line)
            while (matcher.find()) {
                if (!matcher.group(1).equals(
                        "ts",
                        ignoreCase = true
                    ) && !matcher.group(1).equals("name", ignoreCase = true)
                ) {
                    map!![matcher.group(1)] = line
                }
            }
        }
        scanner.close()
    }

    fun convert(data: Reading): List<String>? {
        val type = data.type.toDataString().toLowerCase()
        val map = this.map!!

        val dataString: String

        dataString = if (map.containsKey(type)) {
            map[type]!!.replace(
                "$" + data.type.toDataString().toLowerCase(),
                data.value.toString()
            )
        } else {
            if (map.containsKey(DataStringConverter.VALUE)) {
                map[DataStringConverter.VALUE]!!.replace(
                    "\$value",
                    data.value.toString()
                )
            } else {
                return null
            }
        }

        return listOf(
            dataString
                .replace("\$ts", System.currentTimeMillis().toString())
                .replace("\$name", data.type.toDataString())
        )
    }
}
